src/metric_find_query.ts

```diff
@@ -16,41 +16,63 @@
 // You should have received a copy of the GNU General Public License
 // along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-import { chain, map as _map, uniq } from 'lodash';
-import { lastValueFrom } from 'rxjs';
-import { map } from 'rxjs/operators';
+import { chain, map as _map } from 'lodash';
 
-import { MetricFindValue, TimeRange } from '@grafana/data';
+import { getDefaultTimeRange, MetricFindValue, TimeRange } from '@grafana/data';
 
 import { PrometheusDatasource } from './datasource';
-import { getTimeSrv } from './services/TimeSrv';
-import { PromQueryRequest } from './types';
+import { getVictoriaMetricsTime } from './language_utils';
+import {
+  PrometheusLabelNamesRegex,
+  PrometheusLabelNamesRegexWithMatch,
+  PrometheusMetricNamesRegex,
+  PrometheusQueryResultRegex,
+} from './migrations/variableMigration';
 
 export default class PrometheusMetricFindQuery {
   range: TimeRange;
 
-  constructor(private datasource: PrometheusDatasource, private query: string) {
+  constructor(
+    private datasource: PrometheusDatasource,
+    private query: string
+  ) {
     this.datasource = datasource;
     this.query = query;
-    this.range = getTimeSrv().timeRange();
+    this.range = getDefaultTimeRange();
   }
 
-  process(): Promise<MetricFindValue[]> {
-    const labelNamesRegex = /^label_names\(\)\s*$/;
-    const labelValuesRegex = /^label_values\((?:(.+),\s*)?([a-zA-Z_\\.][a-zA-Z0-9_\\.]*)\)\s*$/;
-    const metricNamesRegex = /^metrics\((.+)\)\s*$/;
-    const queryResultRegex = /^query_result\((.+)\)\s*$/;
+  process(timeRange: TimeRange): Promise<MetricFindValue[]> {
+    this.range = timeRange;
+    const labelNamesRegex = PrometheusLabelNamesRegex;
+    const labelNamesRegexWithMatch = PrometheusLabelNamesRegexWithMatch;
+    const labelValuesRegex = /^label_values\((?:(.+),\s*)?([a-zA-Z_][a-zA-Z0-9_]*)\)\s*$/;
+    const metricNamesRegex = PrometheusMetricNamesRegex;
+    const queryResultRegex = PrometheusQueryResultRegex;
     const labelNamesQuery = this.query.match(labelNamesRegex);
+    const labelNamesMatchQuery = this.query.match(labelNamesRegexWithMatch);
+
+    if (labelNamesMatchQuery) {
+      const selector = `{__name__=~".*${labelNamesMatchQuery[1]}.*"}`;
+      return this.datasource.languageProvider.getSeriesLabels(selector, []).then((results) =>
+        results.map((result) => ({
+          text: result,
+        }))
+      );
+    }
+
     if (labelNamesQuery) {
-      return this.labelNamesQuery();
+      return this.datasource.getTagKeys({ filters: [], timeRange });
     }
 
     const labelValuesQuery = this.query.match(labelValuesRegex);
     if (labelValuesQuery) {
```

Comment:
i think we should check the length also

Reply:
In case of no matches, the result will be `null`, not an empty array.